hnix-store-core/src/System/Nix/Nar.hs

```diff
@@ -0,0 +1,169 @@
+import qualified Data.Set                   as Set
+import qualified Data.Text                  as T
+import qualified Data.Text.Encoding         as E
+import           GHC.Int                    (Int64)
+
+import           System.Nix.Path
+
+
+data NarEffects (m :: * -> *) = NarEffets {
+    readFile         :: FilePath -> m BSL.ByteString
+  , listDir          :: FilePath -> m [FileSystemObject]
+  , narFromFileBytes :: BSL.ByteString -> m Nar
+  , narFromDirectory :: FilePath -> m Nar
+}
+
+
+-- Directly taken from Eelco thesis
+-- https://nixos.org/%7Eeelco/pubs/phd-thesis.pdf
+
+-- TODO: Should we use rootedPath, validPath rather than FilePath?
+data Nar = Nar { narFile :: FileSystemObject }
+    deriving (Eq, Ord, Show)
+
+data FileSystemObject =
+    Regular IsExecutable BSL.ByteString
+  | Directory (Set.Set (PathName, FileSystemObject))
+  | SymLink BSL.ByteString
+  deriving (Eq, Show)
+
+-- TODO - is this right? How does thesis define ordering of FSOs?
+instance Ord FileSystemObject where
+    compare (Regular _ c1) (Regular _ c2) = compare c1 c2
+    compare (Regular _ _)  _              = GT
+    compare (Directory s1) (Directory s2) = compare s1 s2
+    compare (Directory _)  _              = GT
+    compare (SymLink l1) (SymLink l2)     = compare l1 l2
+
+data IsExecutable = NonExecutable | Executable
+    deriving (Eq, Show)
+
+
+------------------------------------------------------------------------------
+-- | Serialize Nar to lazy ByteString
+putNar :: Nar -> B.Put
+putNar (Nar file) = header <>
+                    parens (putFile file)
+    where
+
+        header   = str "nix-archive-1"
+
+        putFile (Regular isExec contents) =
+               str "type" >> str "regular"
+            >> (if isExec == Executable
+               then str "executable" <> str ""
+               else return ())
+            >> str "contents" <> str contents
+
+        putFile (SymLink target) =
+               str "type" >> str "symlink" >> str "target" >> str target
+
+        putFile (Directory entries) =
+               str "type" <> str "directory"
+            <> foldMap putEntry entries
+
+        putEntry (PathName name, fso) =
+            str "entry" <>
+            parens (str "name" <>
+                    str (BSL.fromStrict $ E.encodeUtf8 name) <>
+                    str "node" <>
+                    parens (putFile fso))
+
+        parens m = str "(" >> m >> str ")"
+
+        str :: BSL.ByteString -> B.Put
+        str t = let len = BSL.length t
+            in int len <> pad t
+
+        int :: Integral a => a -> B.Put
+        int n = B.putInt64le $ fromIntegral n
+
+        pad :: BSL.ByteString -> B.Put
+        pad bs = do
+          B.putLazyByteString bs
+          B.putLazyByteString (BSL.replicate (padLen (BSL.length bs)) '\NUL')
+
+
+------------------------------------------------------------------------------
+-- | Deserialize a Nar from lazy ByteString
+getNar :: B.Get Nar
```

Comment:
Does this let us operate lazily over the Nar? Or do we have to consume the whole bytestring to be able to do anything with the Nar?

Reply:
IIUC, when i get the implementation right, we can go through `Nar` lazily streaming one from network <-> disk, or pulling out a individual entries in constant space. But the implementation right now is definitely not going to achieve that.